package/base-files/files/lib/functions/system.sh

```diff
@@ -1,105 +1,122 @@
 # Copyright (C) 2006-2013 OpenWrt.org
 
 . /lib/functions.sh
-. /usr/share/libubox/jshn.sh
 
 get_mac_binary() {
 	local path="$1"
-	local offset="$2"
+	local offset="${2:-0}"
+	local length="${3:-6}"
 
 	if ! [ -e "$path" ]; then
 		echo "get_mac_binary: file $path not found!" >&2
 		return
 	fi
 
-	hexdump -v -n 6 -s $offset -e '5/1 "%02x:" 1/1 "%02x"' $path 2>/dev/null
+	hexdump -v -s "$offset" -n "$length" -e '6/1 "%02x"' "$path" |
+		macaddr_canonicalize
 }
 
 get_mac_label_dt() {
 	local basepath="/proc/device-tree"
-	local macdevice="$(cat "$basepath/aliases/label-mac-device" 2>/dev/null)"
-	local macaddr
+	local macdevice mac
 
-	[ -n "$macdevice" ] || return
+	macdevice=$(cat "$basepath/aliases/label-mac-device" 2>/dev/null)
 
-	macaddr=$(get_mac_binary "$basepath/$macdevice/mac-address" 0 2>/dev/null)
-	[ -n "$macaddr" ] || macaddr=$(get_mac_binary "$basepath/$macdevice/local-mac-address" 0 2>/dev/null)
+	[ -n "$macdevice" ] || return
 
-	echo $macaddr
+	mac=$(get_mac_binary "$basepath/$macdevice/mac-address")
+	[ -n "$mac" ] &&
+		echo -n "$mac" ||
+		echo -n $(get_mac_binary "$basepath/$macdevice/local-mac-address")
 }
 
 get_mac_label_json() {
 	local cfg="/etc/board.json"
-	local macaddr
+	local mac
 
 	[ -s "$cfg" ] || return
 
+	. /usr/share/libubox/jshn.sh
+
 	json_init
 	json_load "$(cat $cfg)"
 	if json_is_a system object; then
 		json_select system
-			json_get_var macaddr label_macaddr
+			json_get_var mac label_macaddr
 		json_select ..
 	fi
 
-	echo $macaddr
+	echo -n "$mac"
 }
 
 get_mac_label() {
-	local macaddr=$(get_mac_label_dt)
-
-	[ -n "$macaddr" ] || macaddr=$(get_mac_label_json)
+	local mac
 
-	echo $macaddr
+	mac=$(get_mac_label_dt)
+	[ -n "$mac" ] &&
+		echo -n "$mac" ||
+		echo -n "$(get_mac_label_json)"
 }
 
 find_mtd_chardev() {
-	local INDEX=$(find_mtd_index "$1")
-	local PREFIX=/dev/mtd
+	local mtdname="$1"
+	local prefix index
+
+	index=$(find_mtd_index "$mtdname")
+	[ -d /dev/mtd ] && prefix="/dev/mtd/" || prefix="/dev/mtd"
 
-	[ -d /dev/mtd ] && PREFIX=/dev/mtd/
-	echo "${INDEX:+$PREFIX$INDEX}"
+	echo -n "${index:+${prefix}${index}}"
 }
 
 mtd_get_mac_ascii() {
 	local mtdname="$1"
 	local key="$2"
-	local part
-	local mac_dirty
+	local part value prefix suffix
 
 	part=$(find_mtd_part "$mtdname")
 	if [ -z "$part" ]; then
 		echo "mtd_get_mac_ascii: partition $mtdname not found!" >&2
 		return
 	fi
 
-	mac_dirty=$(strings "$part" | sed -n 's/^'"$key"'=//p')
+	value=$(strings "$part" |
+		while read vars; do
+			case "$vars" in
+				*"$key"*) echo -n "$vars" && break ;;
+			esac
+		done)
```

Comment:
This might result a false positive. https://github.com/openwrt/openwrt/pull/9468#issuecomment-1066621738 https://github.com/openwrt/openwrt/pull/9468#issuecomment-1066839862

If we're looking for "addr" and found "Caddr", then we cannot know whether there is a valid "Caddr" key, or "addr" key is prefixed by the CRC. IMO dropping the entire CRC bytes is the only way to solve the problem robustly.

Reply:
ok so after some thinking

 - added automatic and manual way to deal with CRC bytes, or something similar
 - globbing on both sides of "key" is a backup, globbing after only takes priority

Reply:
It is still based on so many assumptions and conditions. It might work for most cases but it opens more holes to corner cases. (e.g. `"$key"*` will also take `ethaddr1` even the input key is `ethaddr`) I hope there's a simpler approach to this...

Reply:
good point, however we are starting to get into cases that are more and more unlikely. 

so let's just have the break condition an exact match, and glob otherwise?

Reply:
Why don't we just drop the CRC bytes and use the strict matching as before? (no globbing)
IMHO some of your changes seem to be unnecessary complicated, personally I'd prefer to keep the change as minimal, except for the shellcheck fix (thanks for taking care of it BTW).

Reply:
when you last commented I had already placed a way to skip the CRC

now the function also prefers a non-globbing match over a globbing match, in other words globbing is a fallback

Reply:
> 

Sorry for the lateness, but not all environment variables start at the beginning of the mtd partition. For example, the RAISECOM MSG1500 you mentioned, its environment variables are divided into two parts, one part starts from u-boot-env@0x0, and the other part starts from u-boot-env@0x8000. Your approach is good, I think for these more special devices, we can use "fw_printenv" or "fw_printsys" directly, as @mans0n once said.

Reply:
BTW have you really tested it on device? Try calling `mtd_get_mac_ascii` with `set -x` and it will be evident what is really going on.

Reply:
I have been testing during writing and after pushing intermittently, I've been doing more serious testing recently compared to a few days ago

Reply:
@DragonBluep that second environment @ 0x8000 has its own CRC as well? is it a different mtd partition too?

Reply:
Okay, I'm sorry for being harsh.
What I wanted to report:
```
root@OpenWrt:/tmp# mtd_get_mac_ascii config ethaddr
root@OpenWrt:/tmp# set -x
root@OpenWrt:/tmp# mtd_get_mac_ascii config ethaddr
+ mtd_get_mac_ascii config ethaddr
...
+ keys='bootcmd=tftp
ethaddr=xx:xx:xx:xx:xx:xx
...
stdout=serial'
+ mac='ethaddr=xx:xx:xx:xx:xx:xx'
+ continue
+ echo -n '=xx:xx:xx:xx:xx:xx'
+ macaddr_canonicalize
+ local 'mac='
+ local 'canon='
+ echo -n
+ tr -d '"'
+ mac=
+ '[' 0 -gt 17 ]
+ '[' -n  ]
+ '[' 0 -ne 17 ]
+ return
root@OpenWrt:/tmp# 
```

> that second environment @ 0x8000 has its own CRC as well?

It seems so. https://github.com/openwrt/openwrt/pull/9468#issuecomment-1066308396

Reply:
so I did some testing with `set -x` (thanks for that tip)

yeah I see the mistake I made, true "exact match" of "key" will never happen because we are looking for "key...value"

however, canonicalize function would filter the "=", it could be that when you tried the ascii function it was with the original canonicalize, or without the char classes I made....or the char class was broken...